planner/core/memtable_predicate_extractor.go

```diff
@@ -184,6 +217,147 @@ func (helper extractHelper) extractAddressAndTypeCols(schema *expression.Schema,
 	return
 }
 
+// extracts the string pattern column, e.g:
+// SELECT * FROM t WHERE c LIKE '%a%'
+// SELECT * FROM t WHERE c LIKE '%a%' AND c REGEXP '.*xxx.*'
+func (helper extractHelper) extractLikePatternCol(
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+	extractColName string,
+) (
+	remained []expression.Expression,
+	patterns []string,
+) {
+	remained = make([]expression.Expression, 0, len(predicates))
+	extractCols := make(map[int64]*types.FieldName)
+	for i, name := range names {
+		if name.ColName.L == extractColName {
+			extractCols[schema.Columns[i].UniqueID] = name
+		}
+	}
+
+	if len(extractCols) == 0 {
+		return predicates, nil
+	}
+
+	// We use a string array to save multiple patterns because the Golang and Rust don't
+	// support perl-like CNF regular expression: (?=expr1)(?=expr2).
+	// e.g:
+	// SELECT * FROM t WHERE c LIKE '%a%' AND c LIKE '%b%' AND c REGEXP 'gc.*[0-9]{10,20}'
+	for _, expr := range predicates {
+		fn, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			remained = append(remained, expr)
+			continue
+		}
+		var colName string
+		var datums []types.Datum
+
+		switch fn.FuncName.L {
+		case ast.GT, ast.GE, ast.LT, ast.LE, ast.EQ, ast.Like, ast.Regexp:
+			colName, datums = helper.extractColBinaryOpConsExpr(extractCols, fn)
+		}
+		if colName == extractColName {
+			switch fn.FuncName.L {
+			case ast.EQ:
+				patterns = append(patterns, "^"+regexp.QuoteMeta(datums[0].GetString())+"$")
+			case ast.Like:
+				patterns = append(patterns, stringutil.CompileLike2Regexp(datums[0].GetString()))
+			case ast.Regexp:
+				patterns = append(patterns, datums[0].GetString())
+			default:
+				remained = append(remained, expr)
+			}
+		} else {
+			remained = append(remained, expr)
+		}
+	}
+	return
+}
+
+// extracts the time range column, e.g:
+// SELECT * FROM t WHERE time='2019-10-10 10:10:10'
+// SELECT * FROM t WHERE time>'2019-10-10 10:10:10' AND time<'2019-10-11 10:10:10'
+func (helper extractHelper) extractTimeRange(
+	ctx sessionctx.Context,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+	extractColName string,
+) (
+	remained []expression.Expression,
+	// unix timestamp in millisecond
+	startTime int64,
+	endTime int64,
+) {
+	remained = make([]expression.Expression, 0, len(predicates))
+	extractCols := make(map[int64]*types.FieldName)
```

Comment:
How about extract line#295 ~ line #300 to a method? since every extractor function will use this logic.